### ISSUE TITLE
UOT: how to output transport matrix larger than GPU memory? 

### ISSUE THREAD

**Issue by feiyang-k** [ott-jax/ott#427]
Hi!

I'm very interested in computing unbalanced OT problems, where the total mass of one distribution can be significantly smaller than the other. 

In this case, I wish to get the following result: which samples in the larger distribution are mapped to?

There seems to be two way to get this info. 

- In dual solutions, the samples that are not mapped should have a zero value as they are not in any transport plans. 
- In primal solutions, the transport map could provide the direct answer to the mapping between samples.

Super interestingly, I did not get the intended results from either approach. For the first approach, most dual solutions have a value away from zero. This is expected because the numeric solutions are only approximate. But I do expect to see the values are different–a group of values near zero and a group of values away from zero. What I got is more of a skewed unimodal distribution and with a small singular spike near zero. In this case, I am totally unable to separate the values that should be zero or nonzero. 

For the second approach, the transport matrix seems larger than the GPU memory. Thus I will run into error when I try to output `ot.matrix`. Is there a method to do this computation in CPU memory?

Besides, `ot.transport_mass` gives the following output 
```python
<bound method SinkhornOutput.cost_at_geom of SinkhornOutput(f=DeviceArray([-8.9249543e-06,  7.2023536e-06, -3.8150220e-06, ...,
              1.0135882e-05,  4.4986818e-06,  8.8782713e-07],            dtype=float32), g=DeviceArray([-0.01647191, -0.01978394, -0.05065671, ..., 10.788136  ,
             14.480527  , 15.55767   ], dtype=float32), errors=DeviceArray([ 4.8087921e-04, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
...
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00, -1.0000000e+00,
             -1.0000000e+00, -1.0000000e+00], dtype=float32), reg_ot_cost=DeviceArray(17.086058, dtype=float32), ot_prob=<ott.core.linear_problems.LinearProblem object at 0x7f9c9c4e0450>)>
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...
```

Is this what it is supposed to be? How should I interpret this?

Many thanks!

**Comment by marcocuturi** [comment]
Thanks for the great questions!
> In dual solutions, the samples that are not mapped should have a zero value as they are not in any transport plans.

I think this is a misunderstanding. For instance, in Sinkhorn, if `f_i=g_j=0`, then it just means the transported mass is `exp(-c_ij/ epsilon)`. So it's rather when both `f_i, g_j << 0` that you could make that statement.
 
> In primal solutions, the transport map could provide the direct answer to the mapping between samples.

yes, correct.

> For the second approach, the transport matrix seems larger than the GPU memory. Thus I will run into error when I try to output ot.matrix. Is there a method to do this computation in CPU memory?

For very large scale problems, you might try to instantiate the transport matrix one line at a time. Rather than writing `ot.matrix`, do instead `ot.apply([1,0,0,...])` to get the first line (change the axis for the first column), etc. It's not really fast, but this might be the only way it you are using, for instance, a `grid` geometry, or if you need to use the `batch_size` option. Another approach could be to use a subset of indices, I think @michalk8 had investigated that in the past.

> Is this what it is supposed to be? How should I interpret this?

there might be a a bug, but I don't see immediately why `cost_at_geom` is mentioned in your trace. Can you reproduce the bug with slightly more detail?

**Comment by michalk8** [comment]
> For the second approach, the transport matrix seems larger than the GPU memory. Thus I will run into error when I try to output ot.matrix. Is there a method to do this computation in CPU memory?


Yes, this can be done using [jax.device_put](https://jax.readthedocs.io/en/latest/_autosummary/jax.device_put.html) as:
```python
import jax

devices = jax.devices()
cpu_device = devices[<INDEX_OF_YOUR_CPU_DEVICE>]

prob = ...
solver = ...
out = solver(prob)
out_cpu = jax.device_put(out, cpu_device)
```

> Besides, ot.transport_mass gives the following output

In the older versions, `transport_mass` was a function, in the newer versions it's a property; would suggest updating `ott-jax`.

> For very large scale problems, you might try to instantiate the transport matrix one line at a time. Rather than writing ot.matrix, do instead ot.apply([1,0,0,...]) to get the first line (change the axis for the first column), etc. It's not really fast, but this might be the only way it you are using, for instance, a grid geometry, or if you need to use the batch_size option. Another approach could be to use a subset of indices, I think @michalk8 had investigated that in the past.

Currently, the best solution to materialize the matrix in a batched fashion would be what @marcocuturi suggests, using the `apply` method. We briefly discuss in #234 the subsetting aproach, can look into it soon!
